reference-implementation/lib/abstract-ops/readable-streams.js

```diff
@@ -422,6 +451,230 @@ function ReadableStreamTee(stream, cloneForBranch2) {
+        // successful synchronously-available reads get ahead of asynchronously-available errors.
+        queueMicrotask(() => {
+          reading = false;
+
+          const chunk1 = chunk;
+          let chunk2 = chunk;
+          if (canceled1 === false && canceled2 === false) {
+            chunk2 = CloneArrayBufferView(chunk);
+          }
+
+          if (canceled1 === false) {
+            ReadableByteStreamControllerEnqueue(branch1._controller, chunk1);
+          }
+          if (canceled2 === false) {
+            ReadableByteStreamControllerEnqueue(branch2._controller, chunk2);
+          }
+        });
+      },
+      closeSteps: () => {
+        reading = false;
+        if (canceled1 === false) {
+          ReadableByteStreamControllerClose(branch1._controller);
+        }
+        if (canceled2 === false) {
+          ReadableByteStreamControllerClose(branch2._controller);
+        }
+        if (branch1._controller._pendingPullIntos.length > 0) {
+          ReadableByteStreamControllerRespond(branch1._controller, 0);
+        }
+        if (branch2._controller._pendingPullIntos.length > 0) {
+          ReadableByteStreamControllerRespond(branch2._controller, 0);
+        }
+        resolvePromise(cancelPromise, undefined);
+      },
+      errorSteps: () => {
+        reading = false;
+      }
+    };
+    ReadableStreamDefaultReaderRead(reader, readRequest);
+  }
+
+  function pullWithBYOBReader(view, forBranch2) {
+    if (ReadableStreamDefaultReader.isImpl(reader)) {
+      assert(reader._readRequests.length === 0);
+      ReadableStreamReaderGenericRelease(reader);
+
+      reader = AcquireReadableStreamBYOBReader(stream);
+      forwardReaderError(reader);
+    }
+
+    const readIntoRequest = {
+      chunkSteps: chunk => {
+        // This needs to be delayed a microtask because it takes at least a microtask to detect errors (using
+        // reader._closedPromise below), and we want errors in stream to error both branches immediately. We cannot let
+        // successful synchronously-available reads get ahead of asynchronously-available errors.
+        queueMicrotask(() => {
+          reading = false;
+
+          if (forBranch2 === true) {
+            if (canceled1 === false) {
+              const clonedChunk = CloneArrayBufferView(chunk);
+              ReadableByteStreamControllerEnqueue(branch1._controller, clonedChunk);
+            }
+            if (canceled2 === true) {
+              chunk = new Uint8Array(chunk.buffer, chunk.byteOffset, 0);
+            }
+            ReadableByteStreamControllerRespondWithNewView(branch2._controller, chunk);
+          } else {
+            if (canceled2 === false) {
+              const clonedChunk = CloneArrayBufferView(chunk);
+              ReadableByteStreamControllerEnqueue(branch2._controller, clonedChunk);
+            }
+            if (canceled1 === true) {
+              chunk = new Uint8Array(chunk.buffer, chunk.byteOffset, 0);
```

Comment:
I've gotten lost trying to figure out the purpose of this wrapper... I kind of thought the stream machinery handled the case of returning the buffers when closed for us?

Reply:
This is a tricky one! This is how we get to this point:
1. _branch1_ pulls with a BYOB request. We call _pullWithBYOBReader_ with `forBranch2 = false`, and start a BYOB read on the original stream using the buffer from _branch1_'s BYOB request. Note that this **transfers control of the buffer** from _branch1_ to the original stream.
2. _branch1_ gets canceled. We set `canceled1 = true` in _cancel1Algorithm_. However, we cannot yet "finish" _branch1_'s BYOB request at this point, because we cannot yet return control of the BYOB request's buffer back to _branch1_. Any attempt to call `byobRequest.respond()` here would throw an "already transferred" error (at step 1 in [ReadableByteStreamControllerRespondInClosedState](https://streams.spec.whatwg.org/commit-snapshots/dd76a17a3738d78708a8dfd8f0508e717d6a1988/#readable-byte-stream-controller-respond-in-closed-state)).
3. The BYOB read on the original stream completes. We get passed _chunk_, filled with new data. But more importantly: we get control back over _chunk_'s underlying buffer! So **now** we can finally respond to _branch1_'s BYOB request. But we can't use `byobRequest.respond()`, since that will still try to use the old (transferred) buffer. We **have** to use `.respondWithNewView()` instead, so it'll use the new buffer.
4. Since `canceled1 === true`, `branch1.[[state]] === "closed"`. This means we **must** respond to the BYOB request with exactly 0 bytes written, as per step 3.1 in [ReadableByteStreamControllerRespondInternal](https://streams.spec.whatwg.org/commit-snapshots/dd76a17a3738d78708a8dfd8f0508e717d6a1988/#readable-byte-stream-controller-respond-internal). But the chunk *was* filled with some data (which we can still copy and pass to _branch2_), so we first have to turn the filled view into an empty one before responding.

Reply:
> I kind of thought the stream machinery handled the case of returning the buffers when closed for us?

No, it doesn't. If you look at [ReadableStreamClose](https://streams.spec.whatwg.org/commit-snapshots/dd76a17a3738d78708a8dfd8f0508e717d6a1988/#readable-stream-close), it only runs the close steps of read requests from a *default reader*. It doesn't run any close steps of read-into requests from a *BYOB reader*.

And that's because *it cannot*. In order to run those close steps, it must construct a _chunk_ with a transferred version of the original read-into request's buffer. But it doesn't *have* that buffer, since the underlying byte source might have transferred it from `byobRequest.view` to some other API to do the reading. (For example, by passing it to another `byobReader.read(view)` call like in `ReadableByteStreamTee`.)

We can only call these close steps when we have received control back of that buffer, which must happen through either `byobRequest.respond()` (if the byte source didn't transfer it and instead wrote directly into the buffer) or `.respondWithNewView()` (if it did transfer it). 

This also means that *every underlying byte source implementation* must do something like this in its `cancel` algorithm:
```javascript
new ReadableStream({
  type: 'bytes',
  start(controller) {
    this._controller = controller;
  },
  pull(controller) {
    // call controller.enqueue() or controller.byobRequest.respond() somehow...
  },
  cancel() {
    // finish any pending BYOB request (even if we never used it ourselves)
    this._controller.byobRequest?.respond(0);
  }
});
```
Otherwise, the last BYOB read can never resolve to `{ value: emptyChunk, done: true }` after canceling the stream.

Reply:
Thanks so much for the analysis and walking me through it!

> This also means that every underlying byte source implementation must do something like this in its cancel algorithm:

This seems like a significant footgun, especially given that none of the [example byte sources in the spec](https://streams.spec.whatwg.org/#example-rbs-push) do so.

Can we make this easier somehow? E.g.:

- Automatically introspect `byobRequest.view` and if it's transferred, call (the equivalent of) `respondWithNewView(constructedView)`, and if it's not transferred, call `respond(0)`?

- Relax the requirement that the last BYOB read should resolve with `value: emptyChunk`?

Reply:
> This seems like a significant footgun, especially given that none of the [example byte sources in the spec](https://streams.spec.whatwg.org/#example-rbs-push) do so.

Yes, it really is. We'll need to update those to handle it properly.

> Can we make this easier somehow?
> * Automatically introspect `byobRequest.view` and if it's transferred, call (the equivalent of) `respondWithNewView(constructedView)` ...

I don't think that's possible? If a buffer is transferred, we have no way to know *where it was transferred to*. We definitely cannot re-construct an `ArrayBuffer` around something that has been transferred. For all we know, that buffer is now being used by a different thread in a Web Worker.

Until the underlying byte source actually calls `.respond()` or `.respondWithNewView()` itself, we have no way to "reclaim" the transferred buffer.

> * ...and if it's not transferred, call `respond(0)`?

That *might* work, but it can cause unexpected behavior:
```javascript
const rs = new ReadableStream({
  type: "bytes",
  pull(controller) {
    const byobRequest = controller.byobRequest;
    if (byobRequest) {
      // Note: we're not returning this promise, so the stream will not wait for it before canceling.
      fetchSomeBytes(byobRequest.view.byteLength).then((data) => {
        // If the stream is canceled while fetching the data, then the stream would have already "auto-responded"
        // to the BYOB request. This invalidates the request: its buffer is transferred back to the stream,
        // and its view is set to null.
        byobRequest.view.set(data, 0); // 💥
        byobRequest.respond(data.byteLength);
      });
    }
  }
}
```

> * Relax the requirement that the last BYOB read should resolve with `value: emptyChunk`?

This would break [the `readInto()` example](https://streams.spec.whatwg.org/#example-manual-read-bytes). The BYOB reader *needs* to get its buffer back.

Reply:
> That might work, but it can cause unexpected behavior:

Eh, seems OK. Your fault for not returning the promise to signal you were still planning to use `byobRequest`.

> This would break the readInto() example. The BYOB reader needs to get its buffer back.

Well, the readInto() example doesn't cancel the stream. I guess maybe that's my suggestion (and perhaps even the current intent of the spec): if you cancel the stream with a BYOB request pending, you don't expect to get your buffer back. You should have waited to cancel until the outstanding BYOB request completed. How bad would that be?

Reply:
> Eh, seems OK. Your fault for not returning the promise to signal you were still planning to use `byobRequest`.

Huh, hang on. We don't wait until `source.pull()` resolves/rejects before we call `source.cancel()`. So even *if* you returned a promise from `pull()`, you'd still have to provide your own `cancel()` method which ensures that the previous `pull()` promise stops early and doesn't use `byobRequest` anymore *after* `cancel()` settles.

I guess that makes sense. We have the same thing for `ReadableByteStreamController.enqueue()`: if the stream's state is `"closed"`, it throws a `TypeError`. So if the `pull()` method still uses it after the stream has been cancelled, it'll (hopefully) propagate the error to the returned promise and the stream will ignore that promise's result (because the state is no longer `"readable"`).

> Well, the readInto() example doesn't cancel the stream. I guess maybe that's my suggestion (and perhaps even the current intent of the spec): if you cancel the stream with a BYOB request pending, you don't expect to get your buffer back. You should have waited to cancel until the outstanding BYOB request completed. How bad would that be?

Hmm, that's also true. That would change `ReadableStreamBYOBReadResult` slightly, since `value` can now be `undefined` if the read resolves as a result of canceling the stream.

But really, what *would* a developer still want to do with the buffer from the `read()` result after they've cancelled the stream? Is there ever a case where you start a BYOB read, then change your mind and cancel it *but still want your buffer back*? 🤔

I'll try it out on the reference implementation, see where it leads us. But I'm *definitely* going to move it to a separate PR now! 😁

Reply:
Solved in #1129. Since we no longer need to respond after the branch was cancelled, I've removed this bit. 🙂